cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -939,15 +939,11 @@ impl MachInstEmit for Inst {
                 assert_eq!(kind.rs2, zero_reg());
                 match taken {
                     BranchTarget::Label(label) => {
-                        // WARNING: next two put_string MUST be in that order,
-                        // otherwise, if kind.rs1 == A it will lead to infinite loops.
-                        // Idk how to make kind.rs1 always not equal A.
-                        // At least, both A and B marked as clobbered in
-                        // mod.rs, so it will work.
-                        put_string(&format!("{} => B\n", reg_name(kind.rs1)), sink);
-                        put_string("0 => A\n", sink);
-                        put_string("$ => A :EQ\n", sink);
-                        put_string(&format!("A :JMPZ(label_{})\n", label.index()), sink);
+                        // Due to ISLE rules for `cond_br` the comparison result is in `kind.rs1`.
```

Comment:
I think it's worth documenting that we rely here on assumptions that `rs2` contains zero and that the opcode is `NotEquals`. If any of these is violated, this code will become incorrect.

Reply:
The correct thing to do would be to either:

* Make `CondBr` not rely on this;
* Change `CondBr` to something like `CondBr { condition: ..., target: ..., rs1: ... }` which would no longer imply that it compares stuff internally.

I think either option can be a good one for zkASM specifically (the latter option may potentially allow for more optimization in very specific cases, the former option allows to take more advantage of ISLE in combining), but in either case this should become a `TODO` rather than us spending effort to carefully document a poorly constructed abstraction.

Reply:
> I think it's worth documenting that we rely here on assumptions that `rs2` contains zero and that the opcode is `NotEquals`. If any of these is violated, this code will become incorrect.

I can’t follow why these assumptions are needed? My understanding of [these ISLE rules](https://github.com/near/wasmtime/blob/ce6793c087ccf2c093c3271ea0cfa4a738c9af8a/cranelift/codegen/src/isa/zkasm/inst.isle#L1534-L1568) is that `CondBr` already receives the result of the comparison in `rs1`, which is why we can use it directly for `:JMPNZ`.

Reply:
I've created #179 and in 786ad1a added a related TODO to the code and extended the comment.